File: event-service/src/main/java/renovation/event/service/service/producer/KafkaProducer.java
package renovation.event.service.service.producer;

public interface KafkaProducer<K, V> {

    /**
     * Send data to Kafka broker(s) into default topic.
     *
     * @param data data for to be sent in Kafka
     */
    void send(V data);

    /**
     * Send data to Kafka broker(s) into default topic.
     *
     * @param key  key for to be sent in Kafka
     * @param data data for to be sent in Kafka
     */
    void send(K key, V data);

    /**
     * Send data to Kafka broker(s) into specified topic.
     *
     * @param key  key for to be sent in Kafka
     * @param data data for to be sent in Kafka
     * @param specifiedTopic topic for to be sent in Kafka
     */
    void send(K key, V data, String specifiedTopic);
}

File: kafka-service/src/main/kotlin/renovation/kafka/service/simple/KafkaProducer.kt
/*
 * Created under not commercial project "Renovation"
 *
 * Copyright 2021-2023
 */

package renovation.kafka.service.simple

import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import org.springframework.stereotype.Component

@Configuration
@Profile("simple")
class KafkaProducerConfig {
    @Bean
    fun producerFactory(@Value("\${spring.kafka.bootstrap-servers}") kafkaServerUrl: String):
        ProducerFactory<String, String> {
        val configProps: MutableMap<String, Any> = HashMap()
        configProps[ProducerConfig.BOOTSTRAP_SERVERS_CONFIG] = kafkaServerUrl
        configProps[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java
        configProps[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java
        return DefaultKafkaProducerFactory(configProps)
    }

    @Bean
    fun kafkaTemplate(producerFactory: ProducerFactory<String, String>) = KafkaTemplate(producerFactory)
}

@Component
@Profile("simple")
class MessageProducer(val kafkaTemplate: KafkaTemplate<String, String>) {

    fun sendMessage(topic: String, message: String?) {
        kafkaTemplate.send(topic, message)
    }
}
